subsys/bluetooth/audio/ascs.c

```diff
@@ -954,23 +953,6 @@ static void disconnected(struct bt_conn *conn, uint8_t reason)
 		if (ase->ep.status.state != BT_BAP_EP_STATE_IDLE) {
 			ase_release(ase);
 			/* At this point, `ase` object have been free'd */
-
-			if (stream != NULL) {
-				const struct bt_bap_stream_ops *ops;
-
-				/* Notify upper layer */
-				ops = stream->ops;
-				if (ops != NULL && ops->released != NULL) {
-					ops->released(stream);
-				} else {
-					LOG_WRN("No callback for released set");
-				}
-			}
```

Comment:
I might be missing something, but removing this may not be correct. 

As far as I can tell, the flow is

1. ACL Disconnects -> `disconnected`
2. `disconnected` -> `ase_release` if not `IDLE`
3. `ase_release` -> `ascs_ep_set_state(&ase->ep, BT_BAP_EP_STATE_RELEASING);`
4. `ascs_ep_set_state(&ase->ep, BT_BAP_EP_STATE_RELEASING);` -> `ascs_ep_set_state(ep, BT_BAP_EP_STATE_IDLE);` if `BT_ISO_STATE_DISCONNECTED` else `ascs_disconnect_stream`

In other words, `ops->released` will only be called here if the CIS disconnects before the ACL (which is not a given). 

Your assumption (https://github.com/zephyrproject-rtos/zephyr/pull/57835/files#r1193347864) that `ase_release` will always call `ascs_ep_set_state(BT_BAP_EP_STATE_IDLE)` is incorrect

Reply:
You're right, my comment (https://github.com/zephyrproject-rtos/zephyr/pull/57835/files#r1193347864) is a bit simplified, however the `ascs_ep_set_state(BT_BAP_EP_STATE_IDLE)` will be called eventually anyway.
Let me complete your flow:

> 1. ACL Disconnects -> `disconnected`
> 2. `disconnected` -> `ase_release` if not `IDLE`
> 3. `ase_release` -> `ascs_ep_set_state(&ase->ep, BT_BAP_EP_STATE_RELEASING);`
4a. if `BT_ISO_STATE_DISCONNECTED`: `ascs_ep_set_state(ep, BT_BAP_EP_STATE_IDLE);`
4b. else `ascs_disconnect_stream`
5b. `ascs_iso_disconnected` -> `ascs_ep_set_state(ep, BT_BAP_EP_STATE_IDLE);` if BT_BAP_EP_STATE_RELEASING

Reply:
@Thalley I've added unit test for this particular case `test_release_ase_on_acl_disconnection_server_terminates_cis`

Reply:
> ascs_ep_set_state(BT_BAP_EP_STATE_IDLE) will be called eventually anyway.

Agreed, but the difference is whether or not the stream object is reusable right after the application gets the ACL disconnect callback. 

The code to release the stream here was added for that purpose, and this would have broken then shell earlier as it expected the streams to be released at ACL disconnect. 

Now that I think about it, since we do not know the order of the app vs ASCS ACL disconnect callback, the application can still not expect the streams to be released by the time the callback gets called.

I am OK with this change, and I see that the shell no longer triggers the unicast group delete on ACL disconnect, but rather on stream released.